src/database-controller/src/watcher/framework/index.js

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+require('module-alias/register');
+require('dotenv').config();
+const fetch = require('node-fetch');
+const AsyncLock = require('async-lock');
+const logger = require('@dbc/core/logger');
+const { getFrameworkInformer } = require('@dbc/core/k8s');
+const { alwaysRetryDecorator } = require('@dbc/core/util');
+const config = require('@dbc/watcher/framework/config');
+
+const lock = new AsyncLock({ maxPending: Number.MAX_SAFE_INTEGER });
+
+async function synchronizeFramework(eventType, apiObject) {
+  const res = await fetch(
+    `${config.writeMergerUrl}/api/v1/watchEvents/${eventType}`,
+    {
+      method: 'POST',
+      body: JSON.stringify(apiObject),
+      headers: { 'Content-Type': 'application/json' },
+      timeout: config.writeMergerConnectionTimeoutSecond * 1000,
+    },
+  );
+  if (!res.ok) {
+    throw new Error(`Request returns a ${res.status} error.`);
+  }
+}
+
+const eventHandler = (eventType, apiObject) => {
+  /*
+    framework name-based lock + always retry
+  */
+  const receivedTs = new Date().getTime();
+  const state =
+    apiObject.status && apiObject.status.state
+      ? apiObject.status.state
+      : 'Unknown';
+  logger.info(
+    `Event type=${eventType} receivedTs=${receivedTs} framework=${apiObject.metadata.name} state=${state} received.`,
+  );
+  lock.acquire(
+    apiObject.metadata.name,
+    alwaysRetryDecorator(
+      () => synchronizeFramework(eventType, apiObject),
+      `Sync to write merger type=${eventType} receivedTs=${receivedTs} framework=${apiObject.metadata.name} state=${state}`,
+    ),
+  );
+};
+
+const informer = getFrameworkInformer();
+
+informer.on('add', apiObject => {
+  eventHandler('ADDED', apiObject);
+});
+informer.on('update', apiObject => {
+  eventHandler('MODIFED', apiObject);
+});
+informer.on('delete', apiObject => {
+  eventHandler('DELETED', apiObject);
+});
+informer.on('error', err => {
+  // If any error happens, the process should exit, and let Kubernetes restart it.
+  logger.error(err);
+  process.exit(1);
+});
```

Comment:
Why not inplace restart only the informer?

What is the offical suggested way to handle err?

Reply:
Will network connection drop cause this err?

---
In reply to: [466199663](https://github.com/microsoft/pai/pull/4761#discussion_r466199663) [](ancestors = 466199663)

Reply:
If the connection closes normally, the informer will auto-restart a new connection. Function chain:  
https://github.com/kubernetes-client/javascript/blob/24aec7725aa5fd98c73ce52251d9b09780976dc6/src/cache.ts#L33
https://github.com/kubernetes-client/javascript/blob/24aec7725aa5fd98c73ce52251d9b09780976dc6/src/cache.ts#L92
https://github.com/kubernetes-client/javascript/blob/79736b9a608c18d818de61a6b44503a08ea3a78f/src/watch.ts#L73

If the connection closes unexpectedly, the informer will throw an error.

I don't want to handle this error because it will involve logic like backoff/retrying. Using k8s's backoff strategy brings convenience.

Reply:
What do you mean by "closes normally or unexpectedly"?  Can you give some examples?
We may anyway to improve this..

Reply:
e.g. 1. cut the cable or disable the network device  2. sudden power failure 3. the process is killed without a graceful shutdown